### PR TITLE
Converting all Search tests to boost

### PR DESCRIPTION
**Pull request by aprokop**
Remains to be done:
- ~Combine `tt::per_element()` with `tt::tolerance(1e-14)`~
  Luckily, we have exact matches for LinearBVH test, and never test it for distributed (though, we should!)
- [optional] Be able to compare tuples in BOOST_TEST
  Right now fails with `error: static assertion failed with "Type has to implement operator<< to be printable"`
  The workaround is trivial: do individual checks through `get<i>(a) == get<i>(b)`.

### CODE REVIEW

File: packages/Search/test/DTK_EnableViewComparison.hpp [hunk 1]
```diff
@@ -39,6 +39,26 @@ struct bt_iterator_traits<Kokkos::View<T, P...>, true>
...
+    : public boost::mpl::true_
+{
+};
+
+template <typename T, size_t N, typename Proxy>
+struct bt_iterator_traits<Kokkos::Array<T, N, Proxy>, true>
+{
+    using array_type = Kokkos::Array<T, N, Proxy>;
+    using value_type = typename array_type::value_type;
+    using const_iterator = typename std::add_pointer<const value_type>::type;
```
Comment (aprokop) [inline packages/Search/test/DTK_EnableViewComparison.hpp]: Not sure if this is the best way to do.

File: packages/Search/test/Search_UnitTestHelpers.hpp [hunk 2]
```diff
@@ -69,20 +68,18 @@ void checkResults( DataTransferKit::BVH<DeviceType> const &bvh,
     auto distances_host = Kokkos::create_mirror_view( distances );
     deep_copy( distances_host, distances );
 
-    TEST_COMPARE_ARRAYS( indices_host, indices_ref );
-    TEST_COMPARE_ARRAYS( offset_host, offset_ref );
-    TEST_COMPARE_FLOATING_ARRAYS( distances_host, distances_ref, 1e-14 );
+    BOOST_TEST( indices_host == indices_ref, tt::per_element() );
+    BOOST_TEST( offset_host == offset_ref, tt::per_element() );
+    // BOOST_TEST( distances_host == distances_ref, tt::per_element(),
+    // tt::tolerance( 1e-14 ) ); FIXME_BOOST
```
Comment (aprokop) [inline packages/Search/test/Search_UnitTestHelpers.hpp]: This needs to be either worked around, or fixed.

File: packages/Search/test/Search_UnitTestHelpers.hpp [hunk 3]
```diff
@@ -140,10 +141,11 @@ void checkResults(
...
 
-    TEST_COMPARE_ARRAYS( indices_host, indices_ref );
-    TEST_COMPARE_ARRAYS( offset_host, offset_ref );
-    TEST_COMPARE_ARRAYS( ranks_host, ranks_ref );
-    TEST_COMPARE_FLOATING_ARRAYS( distances_host, distances_ref, 1e-14 );
+    BOOST_TEST( indices_host == indices_ref, tt::per_element() );
+    BOOST_TEST( offset_host == offset_ref, tt::per_element() );
+    BOOST_TEST( ranks_host == ranks_ref, tt::per_element() );
+    // BOOST_TEST( distances_host == distances_ref, tt::per_element(),
+    // tt::tolerance( 1e-14 ) ); FIXME_BOOST
```
Comment (aprokop) [inline packages/Search/test/Search_UnitTestHelpers.hpp]: This needs to be either worked around, or fixed.

File: packages/Search/test/Search_UnitTestHelpers.hpp [hunk 4]
```diff
@@ -109,11 +106,15 @@ void checkResults(
...
+        BOOST_TEST( l.size() == r.size() );
         int const n = l.size();
-        TEST_EQUALITY( n, offset_ref[i + 1] - offset_ref[i] );
+        BOOST_TEST( n == offset_ref[i + 1] - offset_ref[i] );
         for ( int j = 0; j < n; ++j )
-            TEST_ASSERT( l[j] == r[j] );
+        {
+            // FIXME_BOOST would be nice if we could compare tuples
+            BOOST_TEST( std::get<0>( l[j] ) == std::get<0>( r[j] ) );
+            BOOST_TEST( std::get<1>( l[j] ) == std::get<1>( r[j] ) );
```
Comment (aprokop) [inline packages/Search/test/Search_UnitTestHelpers.hpp]: Just wanted to make a notice of this.

File: packages/Search/test/Search_UnitTestHelpers.hpp (hunk 2 above)
Comment (Rombur) [reply packages/Search/test/Search_UnitTestHelpers.hpp]: There is no real work around, you can only use one argument. What you can do is set the tolerance at the `BOOST_AUTO_CASE` level and keep the `tt::per_element()` at the `BOOST_TEST` level.

File: packages/Search/test/tstDetailsTreeTraversal.cpp [hunk 5]
```diff
@@ -14,235 +14,240 @@
...
     // Will break if data is reordered in the PriorityQueue class declaration.
     auto heap =
         reinterpret_cast<typename PriorityQueue::value_type const *>( &queue );
     for ( typename PriorityQueue::size_type i = 0; i < size; ++i )
-        TEST_EQUALITY( heap[i], heap_ref[i] );
+        BOOST_TEST( heap[i] == heap_ref[i] );
 }
 
-TEUCHOS_UNIT_TEST( HeapOperations, push_heap )
+BOOST_AUTO_TEST_CASE( heap_operations_push_heap )
```
Comment (dalg24) [inline packages/Search/test/tstDetailsTreeTraversal.cpp]:
```suggestion
BOOST_AUTO_TEST_CASE( push_heap )
```

File: packages/Search/test/tstDetailsTreeTraversal.cpp [hunk 6]
```diff
@@ -14,235 +14,240 @@
...
     dtk::pushHeap( a.data(), a.data() + 6, dtk::Less<int>() );
-    TEST_ASSERT( dtk::isHeap( a.data(), a.data() + 6, dtk::Less<int>() ) );
+    BOOST_TEST( dtk::isHeap( a.data(), a.data() + 6, dtk::Less<int>() ) );
     Kokkos::Array<int, 6> ref = {15, 5, 11, 3, 4, 8};
-    TEST_COMPARE_ARRAYS( a, ref );
+    BOOST_TEST( a == ref, tt::per_element() );
 }
 
-TEUCHOS_UNIT_TEST( HeapOperations, pop_heap )
+BOOST_AUTO_TEST_CASE( heap_operations_pop_heap )
```
Comment (dalg24) [inline packages/Search/test/tstDetailsTreeTraversal.cpp]:
```suggestion
BOOST_AUTO_TEST_CASE( pop_heap )
```

File: packages/Search/test/tstDetailsTreeTraversal.cpp [hunk 7]
```diff
@@ -14,235 +14,240 @@
...
-    TEST_ASSERT( !dtk::isHeap( a.data(), a.data() + 5, dtk::Less<int>() ) );
+    BOOST_TEST( dtk::isHeap( a.data(), a.data() + 4, dtk::Less<int>() ) );
+    BOOST_TEST( !dtk::isHeap( a.data(), a.data() + 5, dtk::Less<int>() ) );
     Kokkos::Array<int, 5> ref = {8, 5, 4, 3, 11};
-    TEST_COMPARE_ARRAYS( a, ref );
+    BOOST_TEST( a == ref, tt::per_element() );
 }
 
-TEUCHOS_UNIT_TEST( HeapOperations, max_heap )
+BOOST_AUTO_TEST_CASE( heap_operations_max_heap )
```
Comment (dalg24) [inline packages/Search/test/tstDetailsTreeTraversal.cpp]:
```suggestion
BOOST_AUTO_TEST_CASE( max_heap )
```

File: packages/Search/test/tstDetailsTreeTraversal.cpp [hunk 8]
```diff
@@ -258,46 +263,46 @@ TEUCHOS_UNIT_TEST( HeapOperations, max_heap )
...
+    BOOST_TEST( a == ref, tt::per_element() );
 
     dtk::pushHeap( a.data(), a.data() + 7, dtk::Less<int>() );
     ref = {9, 4, 6, 1, 1, 3, 5};
-    TEST_COMPARE_ARRAYS( a, ref );
+    BOOST_TEST( a == ref, tt::per_element() );
 }
 
-TEUCHOS_UNIT_TEST( HeapOperations, min_heap )
+BOOST_AUTO_TEST_CASE( heap_operations_min_heap )
```
Comment (dalg24) [inline packages/Search/test/tstDetailsTreeTraversal.cpp]:
```suggestion
BOOST_AUTO_TEST_CASE( min_heap )
```

File: packages/Search/test/tstDetailsTreeTraversal.cpp [hunk 9]
```diff
@@ -397,14 +401,14 @@ void check_heap( PriorityQueue const &queue, bool &success,
...
+            BOOST_TEST( !compare( heap[parent], heap[i] ) );
         for ( int child : {2 * i + 1, 2 * i + 2} )
             if ( child < size )
-                TEST_ASSERT( !compare( heap[i], heap[child] ) )
+                BOOST_TEST( !compare( heap[i], heap[child] ) );
     }
 }
 
-TEUCHOS_UNIT_TEST( PriorityQueue, maintain_heap_properties )
+BOOST_AUTO_TEST_CASE( priority_queue_maintain_heap_properties )
```
Comment (dalg24) [inline packages/Search/test/tstDetailsTreeTraversal.cpp]: Move test case within `PriorityQueue` "block" and rename.

File: packages/Search/test/tstDetailsTreeTraversal.cpp [hunk 10]
```diff
@@ -443,18 +447,18 @@ TEUCHOS_UNIT_TEST( PriorityQueue, maintain_heap_properties )
...
     while ( !queue.empty() )
     {
         queue.pop();
-        check_heap( queue, success, out );
+        check_heap( queue );
     }
 }
 
-TEUCHOS_UNIT_TEST( HeapOperations, sort_heap )
+BOOST_AUTO_TEST_CASE( heap_operations_sort_heap )
```
Comment (dalg24) [inline packages/Search/test/tstDetailsTreeTraversal.cpp]:
```suggestion
BOOST_AUTO_TEST_CASE( sort_heap )
```

**Comment by aprokop** [comment]
retest this please

**Pull request merged**